### PR TITLE
Use disjunctive normal forms in CofEnv

### LINKED ISSUE #133: Aggresive optimizers for CofEnv/CofThy
Minimum requirements:

1. It should render `Cof.reduce` completely useless.
2. `[(phi \/ psi); (phi \/ psi); (phi \/ psi)]` should be simplified to just `[(phi \/ psi)]`.

Current design decisions:

1. Disjunctive normal forms?

### PR DESCRIPTION
**Pull request by favonia**
Close #133.

It looks promising, but cannot duplicate the success of `loop^10000` yet.

### CODE REVIEW

**Comment by favonia** [comment]
@jonsterling Could you binary search the bottleneck of `loop^` on your computer again? I wonder how much Ch'i has been recollected.

**Comment by jonsterling** [comment]
I will play around with this tomorrow evening! looks pretty cool

**Comment by favonia** [comment]
I think it got even slightly faster now.

> @jonsterling Could you binary search the bottleneck of `loop` on your computer again? I wonder how much Ch'i has been recollected.

**Comment by favonia** [comment]
- [ ] Apply the Cool opitimizations to userland splitting (Split and TpSplit)
- [ ] Even more aggressive rebasing

File: src/lib/CofThy.ml [hunk 1]
```diff
@@ -1,209 +1,233 @@
...
-    assume @@ disj_envelope alg_thy
-
-  let left_invert_under_cofs ~zero ~seq alg_thy phis cont =
-    Search.left_invert ~zero ~seq ~fast_track:(fun _ x -> x ()) (disj_envelope alg_thy) phis @@ fun alg_thy' ->
-    cont @@ `Consistent alg_thy'
+    | [] -> `Inconsistent
+    | _ -> `Consistent
+
+  (* favonia: this optimization seems to be too costly? *)
+  let rebase_branch cached_branches : t =
```
Comment (jonsterling) [inline src/lib/CofThy.ml:198]: So what does rebase do?

**Comment by jonsterling** [comment]
@favonia It is indeed a little faster now. 30,000 loops in 8.4 seconds on my machine, in comparison to about 10 seconds before.

File: src/lib/CofThy.mli [hunk 2]
```diff
@@ -1,57 +1,66 @@
...
 
-module Algebraic :
+module Disj :
 sig
-  val init : unit -> alg_thy
+  type t = disj_thy
 
-  (** Construct the enveloping disjunctive theory of an algebraic theory. *)
-  val disj_envelope : alg_thy -> disj_thy
+  (** Create an empty theory (uses a benign effect for now). *)
```
Comment (jonsterling) [inline src/lib/CofThy.mli:34]: If this uses a benign effect, how come `init` isn't wrapped in a thunk?

File: src/lib/CofThy.ml [hunk 3]
```diff
@@ -1,209 +1,233 @@
...
+    in
+    let useful eq = cached_branches |> List.exists @@ fun (thy', _) -> not @@ Alg.test_eq thy' eq in
+    cached_branches |> List.map @@ fun (thy', (vars, eqs)) ->
+    thy', (VarSet.diff vars common_vars, List.filter useful eqs)
+
+  let split (thy : t) (cofs : cof list) : cached_branches =
+    match disect_cofibrations cofs with
+    | [] -> []
+    | [vars, []] when VarSet.is_empty vars -> thy
+    | disected_cofs ->
```
Comment (jonsterling) [inline src/lib/CofThy.ml:213]: disect => dissect 😜

File: src/lib/CofThy.ml [hunk 4]
```diff
@@ -1,209 +1,233 @@
...
-    | `Consistent thy -> left_invert' ~zero ~seq ~fast_track thy phis cont
+    | `Inconsistent -> []
+    | `Consistent thy' ->
+      match disect_cofibrations cofs with
+      | [] -> []
+      | [vars, []] when VarSet.is_empty vars -> [`Consistent thy']
+      | disected_cofs ->
+        let cached_branches =
+          drop_useless_branches @@
+          shrink_branches thy' disected_cofs in
```
Comment (jonsterling) [inline src/lib/CofThy.ml:176]: when a line is broken like this, I prefer `in` to appear on its own line below

File: src/lib/CofThy.ml (hunk 1 above)
Comment (favonia) [reply src/lib/CofThy.ml:198]: The idea is that, if there are two branches labeled as `phi1 /\ phi2` and `phi1 /\ phi2'`, then it will try to remove the common part `phi1` from all branch labels. There was an actual rebasing but after refactoring maybe it should be called "refactor".

Comment (jonsterling) [reply src/lib/CofThy.ml:198]: oh neat. what do you think about this optimization? is it useful?

File: src/lib/CofThy.ml (hunk 4 above)
Comment (favonia) [reply src/lib/CofThy.ml:176]: LOL the formatting was due to debugging... I will fix that.

File: src/lib/CofThy.mli (hunk 2 above)
Comment (favonia) [reply src/lib/CofThy.mli:34]: LOL that is some leftover from the old comment.

File: src/lib/CofThy.ml (hunk 1 above)
Comment (favonia) [reply src/lib/CofThy.ml]: It seems to help after switching to Map. Maybe it would also help when using the hash tables because I polished the function so much after the switch. Anyways, it is not changing the semantics, so you can always replace it with the identity function and see what would happen.

Comment (favonia) [reply src/lib/CofThy.ml]: Its benefits probably come from faster dropping of useless branches. I need the branch labels for dropping useless branches, and this function tries to factor out the parts of the labels (for faster future dropping).

**Comment by favonia** [comment]
@jonsterling Finally found another possible small improvement!!!

**Comment by favonia** [comment]
@jonsterling [feature/optimization-freeze] I want this PR to enter the "documentation and clean-ups only" mode. No more material optimizations will be introduced.

**Comment by jonsterling** [comment]
> @jonsterling [feature/optimization-freeze] I want this PR to enter the "documentation and clean-ups only" mode. No more material optimizations will be introduced.

@favonia Perfect, I was just about to suggest this 😄

**Comment by favonia** [comment]
I quickly jotted down some of my Thoughts in `CofThy.ml` and `CofThy.mli`. Feel free to wordsmith the comments or rename functions. @cangiuli @jonsterling

File: src/lib/CofThy.ml [hunk 5]
```diff
@@ -1,209 +1,278 @@
...
+      | [] -> VarSet.empty
+      | (_, (vars, _)) :: branches -> List.fold_left go vars branches
+    in
+    (* The following is an approximation of checking whether some equation is useful.
+     * It does not kill every "useless" cofibration. Here is one example:
+     *
+     * branch 1: r=0
+     * branch 2: r=i, i=0
+     *
+     * r=0 will be factored out, but then a=0 should also be removed. Here is a more
```
Comment (jonsterling) [inline src/lib/CofThy.ml:235]: Where did `a=0` come from?

File: src/lib/CofThy.ml [hunk 6]
```diff
@@ -1,209 +1,278 @@
...
+     * enough to detect them. One could consider more aggressive approaches if [CofThy]
+     * becomes the bottleneck again.
+    *)
+    let useful eq = cached_branches |> List.exists @@ fun (thy', _) -> not @@ Alg.test_eq thy' eq in
+    (* revisit all branches and remove all useless ones identified by the simple criterion above. *)
+    cached_branches |> List.map @@ fun (thy', (vars, eqs)) ->
+    thy', (VarSet.diff vars common_vars, List.filter useful eqs)
+
+  (** [split] incorporate a cofibration context and split an already split theory
+    * into more branches. This is similar to [Alg.split] in the spirit, but differs
```
Comment (jonsterling) [inline src/lib/CofThy.ml:251]: I am having trouble following this explanation; could you rephrase or clarify? Is it saying something like:

> `split` adjoins to a disjunctive theory the disjunction of a list of cofibrations, applying certain optimizations beforehand including [XX,YY].

File: src/lib/CofThy.ml [hunk 7]
```diff
@@ -1,209 +1,278 @@
...
+  (** [assume] is the same as [split] except that it will refactor the branches. *)
+  let assume (thy : t) (cofs : cof list) : t =
+    refactor_branches @@ split thy cofs
+
+  let test_sequent thy cx cof =
+    begin
+      split thy cx |> List.map (fun (thy', _) -> thy')
+    end |> List.for_all @@ fun thy' -> Alg.test thy' cof
+
+  let left_invert ~zero ~seq thy cont =
```
Comment (jonsterling) [inline src/lib/CofThy.ml:273]: I am not familiar enough anymore with the algorithm you have implemented to say whether this is still left inversion. Is it still left inversion, or should it be renamed something like `search`?

Comment (favonia) [reply src/lib/CofThy.ml:273]: In case it helps, the disjunctive (normal?) form is implementing eager left inversion. That is, it precomputes all the branches during left inversion the moment any cofibration is added to the theory. Please rename the function to whatever you like. (Do you still like `search` with this explanation?)

Comment (jonsterling) [reply src/lib/CofThy.ml:273]: Gotcha; so it sounds like you are computing the search tree when things are added, and then this function is just traversing the search tree. Is that right?

File: src/lib/CofThy.ml (hunk 5 above)
Comment (favonia) [reply src/lib/CofThy.ml]: From the sky. (Fixed.)

File: src/lib/CofThy.ml (hunk 7 above)
Comment (favonia) [reply src/lib/CofThy.ml:273]: Yes. All the optimizers are to prune the precomputed trees.

Comment (favonia) [reply src/lib/CofThy.ml:273]: PS: I kept the name only because I did not change its semantics.

**Comment by favonia** [comment]
@jonsterling Should I do s/init/empty/ in all cases if they are not taking `()`?

File: src/lib/CofThy.ml (hunk 6 above)
Comment (favonia) [reply src/lib/CofThy.ml]: No, but I updated the comments.
